File: interviews/lc/08-move-zeros.cpp
#include "lc-utility.h"

void moveZeroes(vector<int>& nums) {
  size_t nextInsertionIdx = 0;
  for (size_t i = 0; i < nums.size(); i++) {
    if (nums[i] != 0) {
      const auto tmp = nums[nextInsertionIdx];
      nums[nextInsertionIdx] = nums[i];
      nums[i] = tmp;
      nextInsertionIdx++;
    }
  }
}

void runTestCase(vector<int> nums) {
  const auto numsStr = vector2str(nums);
  moveZeroes(nums);
  const auto resultStr = vector2str(nums);
  cout << numsStr << " => " << resultStr << endl;
}

int main(int argc, char* argv[]) {
  runTestCase({0, 1, 0, 3, 12});
  runTestCase({0});
  return EXIT_SUCCESS;
}

File: interviews/lc/08-move-zeros.py
from typing import List
import time


def moveZeroes(nums: List[int]) -> None:
    insertionIdx = 0
    for i in range(len(nums)):
        if nums[i] != 0:
            nums[insertionIdx], nums[i] = nums[i], nums[insertionIdx]
            insertionIdx += 1


def runTestCase(nums: List[int]) -> None:
    numsStr = ",".join([str(n) for n in nums])
    start = time.perf_counter_ns()
    moveZeroes(nums)
    end = time.perf_counter_ns()
    resultStr = ",".join([str(n) for n in nums])
    delta = end - start
    print(f"{numsStr} => {resultStr} {delta}ns")


if __name__ == "__main__":
    runTestCase([0, 0, 0, 0, 1])
    runTestCase([0, 0, 0, 1, 2])
    runTestCase([1, 2, 3])
    runTestCase([0, 1, 0, 3, 12])
    runTestCase([0])
